src/passes/OptimizeInstructions.cpp

```diff
@@ -2815,6 +2818,31 @@ struct OptimizeInstructions
     return false;
   }
 
+  // (x << z) op (y << z)    ==>   (x op y) << z,   op = `|`, `&`, `^`
+  // (x >> z) op (y >> z)    ==>   (x op y) >> z,   op = `|`, `&`, `^`
+  // (x >>> z) op (y >>> z)  ==>   (x op y) >>> z,  op = `|`, `&`, `^`
+  Expression* combineBinaryWithShifts(Binary* curr) {
```

Comment:
This seems parallel to the existing `preserveAnd` etc. methods. E.g. `preserveAnd` does
```cpp
(x  < 0) & (y  < 0) ==> (x & y)  < 0
```
while this does e.g.
```cpp
(x >> z) & (y >> z) ==> (x & y) >> z
```
Perhaps this could be shared with that somehow, or called `preserveShift`, or such?

Reply:
It can't be shared with `preserveAnd` or `preserveOr` because both these methods operate with constants on right sides, while `combineBinaryWithShifts` is more general. Also, `combineBinaryWithShifts` include Xor operation. But I can change the name of this rule